CMakeLists.txt

```diff
@@ -180,3 +180,36 @@ if(USE_UPDATER)
   endif()
 
 endif()
+
+# Get the latest commit hash
+execute_process(
+        COMMAND git rev-parse --short HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_COMMIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+if(GIT_COMMIT_HASH)
+  STRING(TOUPPER ${GIT_COMMIT_HASH} GIT_COMMIT_HASH)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "-DGIT_COMMIT_HASH=\"${GIT_COMMIT_HASH}\"")
+endif()
+
+# Get the current working branch
+execute_process(
+        COMMAND git describe --all --dirty --broken
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_BRANCH
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+if(GIT_BRANCH)
+  STRING(REGEX REPLACE "heads/" "" GIT_BRANCH ${GIT_BRANCH})
+
+  if(GIT_BRANCH MATCHES ".*-dirty$")
+    STRING(REGEX REPLACE "-dirty$" " (DIRTY)" GIT_BRANCH ${GIT_BRANCH})
+  endif()
+
+  if(GIT_BRANCH MATCHES ".*-broken")
+    STRING(REGEX REPLACE "-broken" " (BROKEN)" GIT_BRANCH ${GIT_BRANCH})
```

Comment:
What value does showing this level of information about the Git branch give us?

Reply:
Because it tells someone viewing the compiled product whether it was built from the exact Git SHA1 or whether the source code was modified (or the local repository is corrupt) after it was checked out but ***before*** it was compiled - otherwise knowing the SHA1 is not as useful for diagnostic purposes...

Reply:
And it doesn't fail command when git repo is broken :)